app/legacy/dao/notifications_dao.py

```diff
@@ -22,11 +36,137 @@ async def get_notification(id: UUID4) -> Row[Any]:
+        to: str | None,
+        reply_to_text: str,
+        service_id: UUID4,
+        api_key_id: UUID4,
+        template_id: UUID4,
+        template_version: int,
+        billable_units: int,
+        key_type: str,
+    ) -> None:
+        legacy_notifications = metadata_legacy.tables['notifications']
+        try:
+            async with get_write_session_with_context() as session:
+                stmt = insert(legacy_notifications).values(
+                    id=id,
+                    notification_type=notification_type.value,
+                    to=to,
+                    reply_to_text=reply_to_text,
+                    service_id=service_id,
+                    api_key_id=api_key_id,
+                    template_id=template_id,
+                    template_version=template_version,
+                    billable_units=billable_units,
+                    created_at=datetime.now(),
+                    key_type=key_type,
+                    notification_status=NotificationStatus.CREATED,
+                )
+                await session.execute(stmt)
+                await session.commit()
+        except (IntegrityError, DataError) as e:
+            if 'duplicate' in str(e).lower():
+                logger.warning('Duplicate key detected: {}', id)
+                raise RetryableError(log_msg='Duplicate key') from e
+            else:
+                # These are deterministic and will likely fail again
+                logger.exception('Notification insert failed: invalid or unexpected data for id: {}', id)
+                raise NonRetryableError('Notification insert failed: invalid or unexpected data.') from e
+
+        except (OperationalError, InterfaceError, TimeoutError) as e:
+            # Transient DB issues that may succeed on retry
+            logger.warning('Notification insert failed due to a transient database error for id: {}', id)
+            raise RetryableError('Notification insert failed due to a transient database error.') from e
+
+        except SQLAlchemyError as e:
+            logger.exception('Unexpected SQLAlchemy error during notification insert for id: {}', id)
+            raise NonRetryableError('Unexpected SQLAlchemy error during notification insert.') from e
+
+    @staticmethod
+    async def delete_notification(
+        id: UUID4,
+    ) -> None:
+        """Delete a Notification from the legacy database.
+
+        Args:
+            id (UUID4): id of the notification
+
+        Raises:
+            NonRetryableError: If unable to delete the notification
+        """
+
+        @db_retry
+        async def _delete() -> None:
+            async with get_write_session_with_context() as session:
+                legacy_notifications = metadata_legacy.tables['notifications']
+                stmt = delete(legacy_notifications).where(legacy_notifications.c.id == id)
+                await session.execute(stmt)
+                await session.commit()
```

Comment:
I believe we need to catch exceptions here and raise retryable / nonretryable exceptions instead.

Reply:
Fixed, thank you!